package-res/resources/web/pentaho/lang/Filter.js

```diff
@@ -1,42 +1,138 @@
 define([
-  "cdf/lib/Base"
-], function(Base) {
+  "pentaho/lang/Base",
+  "pentaho/data/TableView"
+], function (Base, TableView) {
   "use strict";
 
-  var Node = Base.extend({
-    constructor: function(value) {
+  var AbstractFilter = Base.extend({
+    constructor: function (value) {
       this._value = value;
     },
-    predicate: function(v) {
+
+    predicate: function (row) {
       return false;
+    },
+
+    contains: function (element) {
+      return AbstractFilter.contains(this, element);
+    },
+
+    negation: function () {
+      return AbstractFilter.negation(this);
+    },
+
+    union: function (filter) {
+      return AbstractFilter.union(this, filter);
+    },
+
+    intersection: function (filter) {
+      return AbstractFilter.intersection(this, filter);
+    },
+
+    filter: function (dataTable) {
+      var k, nRows = dataTable.getNumberOfRows();
+      var filteredRows = [];
+
+      for (k = 0; k < nRows; k++) {
+        var row = {
+          dataTable: dataTable,
+          rowIdx: k
+        };
+        var bool = this.predicate(row);
+        if (bool) {
+          filteredRows.push(k);
+        }
+      }
+
+      var dataView = new TableView(dataTable);
+      dataView.setSourceRows(filteredRows);
+      return dataView;
+    }
+  }, {
+    contains: function (filter, dataTable) {
```

Comment:
sure, I was just using the dataTable as a placeholder while we were attempting to understand the contains....